python/cuml/metrics/pairwise_distances.pyx

```diff
@@ -0,0 +1,222 @@
+from cuml.common import (get_cudf_column_ptr, get_dev_array_ptr,
+                         input_to_cuml_array, CumlArray, logger, with_cupy_rmm)
+from cuml.metrics.cluster.utils import prepare_cluster_metric_inputs
+
+cdef extern from "cuml/distance/distance_type.h" namespace "ML::Distance":
+
+    cdef enum DistanceType:
+        EucExpandedL2 "ML::Distance::DistanceType::EucExpandedL2"
+        EucExpandedL2Sqrt "ML::Distance::DistanceType::EucExpandedL2Sqrt"
+        EucExpandedCosine "ML::Distance::DistanceType::EucExpandedCosine"
+        EucUnexpandedL1 "ML::Distance::DistanceType::EucUnexpandedL1"
+        EucUnexpandedL2 "ML::Distance::DistanceType::EucUnexpandedL2"
+        EucUnexpandedL2Sqrt "ML::Distance::DistanceType::EucUnexpandedL2Sqrt"
+
+cdef extern from "cuml/metrics/metrics.hpp" namespace "ML::Metrics":
+    void pairwiseDistance(const cumlHandle &handle, const double *x,
+                          const double *y, double *dist, int m, int n, int k,
+                          DistanceType metric, bool isRowMajor) except +
+    void pairwiseDistance(const cumlHandle &handle, const float *x,
+                          const float *y, float *dist, int m, int n, int k,
+                          DistanceType metric, bool isRowMajor) except +
+
+
+def determine_metric(metric_str):
+
+    # Available options in scikit-learn and their pairs. See
+    # sklearn.metrics.pairwise.PAIRWISE_DISTANCE_FUNCTIONS:
+    # 'cityblock': EucUnexpandedL1
+    # 'cosine': EucExpandedCosine
+    # 'euclidean': EucUnexpandedL2Sqrt
+    # 'haversine': N/A
+    # 'l2': EucUnexpandedL2Sqrt
+    # 'l1': EucUnexpandedL1
+    # 'manhattan': EucUnexpandedL1
+    # 'nan_euclidean': N/A
+    # Note: many are duplicates following this:
+    # https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/pairwise.py#L1321
+
+    if metric_str == 'cityblock':
+        return DistanceType.EucUnexpandedL1
+    elif metric_str == 'cosine':
+        return DistanceType.EucExpandedCosine
+    elif metric_str == 'euclidean':
+        return DistanceType.EucUnexpandedL2Sqrt
+    elif metric_str == 'haversine':
+        raise ValueError(" The metric: '{}', is not supported at this time."
+                         .format(metric_str))
+    elif metric_str == 'l2':
+        return DistanceType.EucUnexpandedL2Sqrt
+    elif metric_str == 'l1':
+        return DistanceType.EucUnexpandedL1
+    elif metric_str == 'manhattan':
+        return DistanceType.EucUnexpandedL1
+    elif metric_str == 'nan_euclidean':
+        raise ValueError(" The metric: '{}', is not supported at this time."
+                         .format(metric_str))
+    else:
+        raise ValueError("Unknown metric: {}".format(metric_str))
+
+
+@with_cupy_rmm
+def pairwise_distances(X, Y=None, metric="euclidean", handle=None,
+                       convert_dtype=True, **kwds):
+    """
+    Compute the distance matrix from a vector array X and optional Y.
```

Comment:
Would be nice to see a quick usage example here.

Reply:
Added an example